File: api/app/chat/session_lifecycle.py
from __future__ import annotations

from collections.abc import Callable
from dataclasses import dataclass
from datetime import UTC, datetime
from uuid import UUID

from app.models import (
    ChatAutosaveStrategy,
    ChatMessageRecord,
    ChatSessionRecord,
    EngramCreateResponse,
    EngramSummary,
    MemoryEngramCreate,
)

from .lifecycle_policy import (
    duplicate_snapshot_exists,
    is_low_value_snapshot_abstract,
    select_retention_prune_ids,
    should_take_interval_snapshot,
    should_take_message_count_snapshot,
)

_GENERIC_SNAPSHOT_ABSTRACTS = {
    "",
    "snapshot from active chat session.",
    "snapshot from active chat session",
    "chat snapshot",
    "session snapshot",
}
_AUTOSAVE_SNAPSHOT_TAGS = [
    "autosave_snapshot",
    "session-lifecycle",
    "chat",
]


@dataclass(frozen=True)
class SessionLifecycleDependencies:
    list_session_linked_engrams: Callable[..., list[EngramSummary]]
    list_chat_messages: Callable[..., list[ChatMessageRecord]]
    count_session_messages_by_role: Callable[..., int]
    delete_session_autosave_engrams: Callable[..., list[UUID]]
    create_engram: Callable[..., EngramCreateResponse]


@dataclass(frozen=True)
class LifecycleMaintenanceResult:
    snapshot_engram_id: UUID | None
    pruned_engram_ids: list[UUID]
    skipped_reason: str | None = None


@dataclass(frozen=True)
class AutosaveSnapshotCreateRequest:
    actor_user_id: UUID
    session: ChatSessionRecord
    existing_snapshots: list[EngramSummary]
    embedding_dim: int


@dataclass(frozen=True)
class SnapshotResolutionRequest:
    actor_user_id: UUID
    session: ChatSessionRecord
    autosave_snapshots: list[EngramSummary]
    now: datetime


@dataclass(frozen=True)
class SnapshotCreationRequest:
    actor_user_id: UUID
    session: ChatSessionRecord
    autosave_snapshots: list[EngramSummary]
    should_create: bool
    skipped_reason: str | None
    embedding_dim: int


def _normalize_spaces(value: str) -> str:
    return " ".join(value.strip().split())


def _truncate_text(value: str, max_chars: int) -> str:
    if len(value) <= max_chars:
        return value
    return value[: max_chars - 3].rstrip() + "..."


def is_generic_snapshot_abstract(value: str) -> bool:
    return _normalize_spaces(value).lower() in _GENERIC_SNAPSHOT_ABSTRACTS


def derive_chat_snapshot_abstract(
    messages: list[ChatMessageRecord], max_chars: int = 320
) -> str:
    for role in ("assistant", "user"):
        for message in reversed(messages):
            if message.role != role:
                continue
            normalized = _normalize_spaces(message.content_text)
            if not normalized:
                continue
            return _truncate_text(normalized, max_chars=max_chars)
    return ""


def transcript_markdown(session: ChatSessionRecord, messages: list[ChatMessageRecord]) -> str:
    lines = [
        f"# Chat Session Snapshot: {session.title}",
        "",
        f"- Session ID: {session.session_id}",
        f"- Provider/Model: {session.provider.value}/{session.model_id}",
        "",
    ]
    for message in messages:
        lines.extend(
            [
                f"## {message.role.upper()} ({message.created_at.isoformat()})",
                message.content_text.strip() or "(empty)",
                "",
            ]
        )
    return "\n".join(lines).strip()


def retrieval_text_from_messages(
    messages: list[ChatMessageRecord], tail_count: int = 8
) -> str:
    tail = messages[-tail_count:]
    return " ".join(item.content_text.strip() for item in tail if item.content_text.strip())


def _is_autosave_snapshot(summary: EngramSummary) -> bool:
    return "autosave_snapshot" in {item.strip().lower() for item in summary.tags}


def _list_autosave_snapshots(
    *,
    actor_user_id: UUID,
    session_id: UUID,
    dependencies: SessionLifecycleDependencies,
    limit: int = 500,
) -> list[EngramSummary]:
    linked = dependencies.list_session_linked_engrams(
        session_id=session_id,
        actor_user_id=actor_user_id,
        limit=limit,
        offset=0,
    )
    return [item for item in linked if _is_autosave_snapshot(item)]


def _create_autosave_snapshot(
    *,
    request: AutosaveSnapshotCreateRequest,
    dependencies: SessionLifecycleDependencies,
) -> UUID | None:
    messages = dependencies.list_chat_messages(
        session_id=request.session.session_id,
        actor_user_id=request.actor_user_id,
        limit=500,
        offset=0,
    )
    if not messages:
        return None

    abstract = derive_chat_snapshot_abstract(messages)
    if is_low_value_snapshot_abstract(abstract):
        return None
    if duplicate_snapshot_exists(abstract=abstract, existing_snapshots=request.existing_snapshots):
        return None

    now = datetime.now(UTC)
    created = dependencies.create_engram(
        payload=MemoryEngramCreate(
            project_id=request.session.project_id,
            thread_id=f"chat-session:{request.session.session_id}:autosave",
            title=f"{request.session.title} Autosave {now.strftime('%Y-%m-%d %H:%M:%S')}",
            abstract=abstract,
            detailed_summary_markdown=transcript_markdown(request.session, messages),
            tags=[
                *_AUTOSAVE_SNAPSHOT_TAGS,
                f"autosave_strategy:{request.session.autosave_strategy.value}",
            ],
            keywords=[
                "autosave",
                "snapshot",
                request.session.provider.value,
                request.session.model_id,
            ],
            visibility_scope=request.session.visibility_scope.value,
            source_session_id=request.session.session_id,
            retrieval_text=retrieval_text_from_messages(messages),
        ),
        embedding_dim=request.embedding_dim,
        owner_user_id=request.actor_user_id,
        enrichment_origin="chat.autosave_snapshot",
    )
    return created.engram_id


def _resolve_snapshot_creation(
    *,
    request: SnapshotResolutionRequest,
    dependencies: SessionLifecycleDependencies,
) -> tuple[bool, str | None]:
    if request.session.autosave_strategy == ChatAutosaveStrategy.interval:
        latest_created_at = (
            request.autosave_snapshots[0].created_at
            if request.autosave_snapshots
            else None
        )
        should_create = should_take_interval_snapshot(
            now=request.now,
            latest_snapshot_created_at=latest_created_at,
            interval_minutes=request.session.autosave_interval_minutes,
        )
        return should_create, None if should_create else "interval_not_elapsed"

    if request.session.autosave_strategy == ChatAutosaveStrategy.message_count:
        assistant_message_count = dependencies.count_session_messages_by_role(
            session_id=request.session.session_id,
            actor_user_id=request.actor_user_id,
            role="assistant",
        )
        should_create = should_take_message_count_snapshot(
            assistant_message_count=assistant_message_count,
            min_messages=request.session.autosave_min_messages,
        )
        return should_create, None if should_create else "message_count_threshold_not_met"

    return False, None


def _maybe_create_snapshot(
    *,
    request: SnapshotCreationRequest,
    dependencies: SessionLifecycleDependencies,
) -> tuple[UUID | None, list[EngramSummary], str | None]:
    if not request.should_create:
        return None, request.autosave_snapshots, request.skipped_reason

    created_snapshot_id = _create_autosave_snapshot(
        request=AutosaveSnapshotCreateRequest(
            actor_user_id=request.actor_user_id,
            session=request.session,
            existing_snapshots=request.autosave_snapshots,
            embedding_dim=request.embedding_dim,
        ),
        dependencies=dependencies,
    )
    if created_snapshot_id is None:
        return None, request.autosave_snapshots, "duplicate_or_low_value_snapshot"

    refreshed_snapshots = _list_autosave_snapshots(
        actor_user_id=request.actor_user_id,
        session_id=request.session.session_id,
        dependencies=dependencies,
    )
    return created_snapshot_id, refreshed_snapshots, request.skipped_reason


def run_session_lifecycle_maintenance(
    *,
    actor_user_id: UUID,
    session: ChatSessionRecord,
    embedding_dim: int,
    dependencies: SessionLifecycleDependencies,
) -> LifecycleMaintenanceResult:
    if not session.autosave_enabled or session.autosave_strategy == ChatAutosaveStrategy.off:
        return LifecycleMaintenanceResult(
            snapshot_engram_id=None,
            pruned_engram_ids=[],
            skipped_reason="autosave_disabled",
        )

    now = datetime.now(UTC)
    autosave_snapshots = _list_autosave_snapshots(
        actor_user_id=actor_user_id,
        session_id=session.session_id,
        dependencies=dependencies,
    )
    should_create, skipped_reason = _resolve_snapshot_creation(
        request=SnapshotResolutionRequest(
            actor_user_id=actor_user_id,
            session=session,
            autosave_snapshots=autosave_snapshots,
            now=now,
        ),
        dependencies=dependencies,
    )
    created_snapshot_id, autosave_snapshots, skipped_reason = _maybe_create_snapshot(
        request=SnapshotCreationRequest(
            actor_user_id=actor_user_id,
            session=session,
            autosave_snapshots=autosave_snapshots,
            should_create=should_create,
            skipped_reason=skipped_reason,
            embedding_dim=embedding_dim,
        ),
        dependencies=dependencies,
    )

    prune_ids = select_retention_prune_ids(
        snapshots=autosave_snapshots,
        retention_days=session.retention_days,
        retention_max_snapshots=session.retention_max_snapshots,
        now=now,
    )
    pruned_ids = dependencies.delete_session_autosave_engrams(
        session_id=session.session_id,
        actor_user_id=actor_user_id,
        engram_ids=prune_ids,
    )

    return LifecycleMaintenanceResult(
        snapshot_engram_id=created_snapshot_id,
        pruned_engram_ids=pruned_ids,
        skipped_reason=skipped_reason,
    )

File: internal/chat/session_lifecycle.go
package chat

import (
	"context"
	"fmt"
	"strings"
	"time"

	"engram/internal/models"

	"github.com/google/uuid"
)

var autosaveSnapshotTags = []string{"autosave_snapshot", "session-lifecycle", "chat"}

// SessionLifecycleDependencies captures storage and engram operations used by lifecycle maintenance.
type SessionLifecycleDependencies struct {
	ListSessionLinkedEngrams     func(ctx context.Context, sessionID uuid.UUID, actorUserID uuid.UUID, limit int, offset int) ([]models.EngramSummary, error)
	ListChatMessages             func(ctx context.Context, sessionID uuid.UUID, actorUserID uuid.UUID, limit int, offset int) ([]models.ChatMessageRecord, error)
	CountSessionMessagesByRole   func(ctx context.Context, sessionID uuid.UUID, actorUserID uuid.UUID, role string) (int, error)
	DeleteSessionAutosaveEngrams func(ctx context.Context, sessionID uuid.UUID, actorUserID uuid.UUID, engramIDs []uuid.UUID) ([]uuid.UUID, error)
	CreateEngram                 func(ctx context.Context, payload models.MemoryEngramCreate, embeddingDim int, ownerUserID uuid.UUID, enrichmentOrigin string) (*models.EngramCreateResponse, error)
}

// LifecycleMaintenanceResult captures snapshot creation and pruning outputs.
type LifecycleMaintenanceResult struct {
	SnapshotEngramID *uuid.UUID
	PrunedEngramIDs  []uuid.UUID
	SkippedReason    *string
}

// SessionLifecycleRunInput captures request inputs for lifecycle maintenance.
type SessionLifecycleRunInput struct {
	ActorUserID  uuid.UUID
	Session      models.ChatSessionRecord
	EmbeddingDim int
	Dependencies SessionLifecycleDependencies
	Now          time.Time
}

type sessionLifecycleRun struct {
	ctx                  context.Context
	actorUserID          uuid.UUID
	session              models.ChatSessionRecord
	embeddingDim         int
	dependencies         SessionLifecycleDependencies
	now                  time.Time
	autosaveSnapshots    []models.EngramSummary
	shouldCreateSnapshot bool
	skippedReason        *string
}

// DeriveChatSnapshotAbstract derives a concise abstract from latest assistant/user messages.
func DeriveChatSnapshotAbstract(messages []models.ChatMessageRecord, maxChars int) string {
	normalizedMaxChars := normalizeMaxChars(maxChars)
	for _, role := range []string{"assistant", "user"} {
		for index := len(messages) - 1; index >= 0; index-- {
			message := messages[index]
			if message.Role != role {
				continue
			}
			normalized := normalizeSpaces(message.ContentText)
			if normalized == "" {
				continue
			}
			return truncateText(normalized, normalizedMaxChars)
		}
	}
	return ""
}

// RunSessionLifecycleMaintenance creates autosave snapshots and applies retention pruning.
func RunSessionLifecycleMaintenance(
	ctx context.Context,
	input SessionLifecycleRunInput,
) (LifecycleMaintenanceResult, error) {
	run := newSessionLifecycleRun(ctx, input)
	if !run.autosaveEnabled() {
		return run.disabledResult(), nil
	}
	if err := run.loadAutosaveSnapshots(); err != nil {
		return LifecycleMaintenanceResult{}, err
	}
	if err := run.resolveSnapshotCreationPlan(); err != nil {
		return LifecycleMaintenanceResult{}, err
	}
	createdSnapshotID, err := run.createSnapshotIfNeeded()
	if err != nil {
		return LifecycleMaintenanceResult{}, err
	}
	prunedIDs, err := run.pruneSnapshots()
	if err != nil {
		return LifecycleMaintenanceResult{}, err
	}
	return LifecycleMaintenanceResult{
		SnapshotEngramID: createdSnapshotID,
		PrunedEngramIDs:  prunedIDs,
		SkippedReason:    run.skippedReason,
	}, nil
}

func newSessionLifecycleRun(ctx context.Context, input SessionLifecycleRunInput) sessionLifecycleRun {
	return sessionLifecycleRun{
		ctx:               ctx,
		actorUserID:       input.ActorUserID,
		session:           input.Session,
		embeddingDim:      input.EmbeddingDim,
		dependencies:      input.Dependencies,
		now:               resolveReferenceTime(input.Now),
		autosaveSnapshots: []models.EngramSummary{},
	}
}

func (run *sessionLifecycleRun) autosaveEnabled() bool {
	return run.session.AutosaveEnabled && run.session.AutosaveStrategy != models.ChatAutosaveStrategyOff
}

func (run *sessionLifecycleRun) disabledResult() LifecycleMaintenanceResult {
	return LifecycleMaintenanceResult{
		SnapshotEngramID: nil,
		PrunedEngramIDs:  []uuid.UUID{},
		SkippedReason:    stringPtr("autosave_disabled"),
	}
}

func (run *sessionLifecycleRun) loadAutosaveSnapshots() error {
	autosaveSnapshots, err := listAutosaveSnapshots(
		run.ctx,
		run.actorUserID,
		run.session.SessionID,
		run.dependencies,
	)
	if err != nil {
		return err
	}
	run.autosaveSnapshots = autosaveSnapshots
	return nil
}

func (run *sessionLifecycleRun) resolveSnapshotCreationPlan() error {
	switch run.session.AutosaveStrategy {
	case models.ChatAutosaveStrategyInterval:
		run.resolveIntervalSnapshotCreationPlan()
		return nil
	case models.ChatAutosaveStrategyMessageCount:
		return run.resolveMessageCountSnapshotCreationPlan()
	default:
		run.shouldCreateSnapshot = false
		run.skippedReason = nil
		return nil
	}
}

func (run *sessionLifecycleRun) resolveIntervalSnapshotCreationPlan() {
	latestCreatedAt := latestSnapshotCreatedAt(run.autosaveSnapshots)
	run.shouldCreateSnapshot = ShouldTakeIntervalSnapshot(
		run.now,
		latestCreatedAt,
		SnapshotIntervalMinutes(run.session.AutosaveIntervalMinutes),
	)
	if run.shouldCreateSnapshot {
		run.skippedReason = nil
		return
	}
	run.skippedReason = stringPtr("interval_not_elapsed")
}

func (run *sessionLifecycleRun) resolveMessageCountSnapshotCreationPlan() error {
	assistantMessageCount, err := run.dependencies.CountSessionMessagesByRole(
		run.ctx,
		run.session.SessionID,
		run.actorUserID,
		"assistant",
	)
	if err != nil {
		return err
	}
	run.shouldCreateSnapshot = ShouldTakeMessageCountSnapshot(
		SnapshotMessageCount(assistantMessageCount),
		SnapshotMessageWindow(run.session.AutosaveMinMessages),
	)
	if run.shouldCreateSnapshot {
		run.skippedReason = nil
		return nil
	}
	run.skippedReason = stringPtr("message_count_threshold_not_met")
	return nil
}

func (run *sessionLifecycleRun) createSnapshotIfNeeded() (*uuid.UUID, error) {
	if !run.shouldCreateSnapshot {
		return nil, nil
	}
	createdSnapshotID, err := run.createAutosaveSnapshot()
	if err != nil {
		return nil, err
	}
	if createdSnapshotID == nil {
		run.skippedReason = stringPtr("duplicate_or_low_value_snapshot")
		return nil, nil
	}
	if err := run.loadAutosaveSnapshots(); err != nil {
		return nil, err
	}
	run.skippedReason = nil
	return createdSnapshotID, nil
}

func (run *sessionLifecycleRun) pruneSnapshots() ([]uuid.UUID, error) {
	pruneIDs := SelectRetentionPruneIDs(
		run.autosaveSnapshots,
		SnapshotRetentionDays(run.session.RetentionDays),
		SnapshotRetentionMaxSnapshots(run.session.RetentionMaxSnapshots),
		run.now,
	)
	return run.dependencies.DeleteSessionAutosaveEngrams(
		run.ctx,
		run.session.SessionID,
		run.actorUserID,
		pruneIDs,
	)
}

func listAutosaveSnapshots(
	ctx context.Context,
	actorUserID uuid.UUID,
	sessionID uuid.UUID,
	dependencies SessionLifecycleDependencies,
) ([]models.EngramSummary, error) {
	linked, err := dependencies.ListSessionLinkedEngrams(ctx, sessionID, actorUserID, 500, 0)
	if err != nil {
		return nil, err
	}
	autosave := make([]models.EngramSummary, 0)
	for _, item := range linked {
		if !isAutosaveSnapshot(item) {
			continue
		}
		autosave = append(autosave, item)
	}
	return autosave, nil
}

func isAutosaveSnapshot(summary models.EngramSummary) bool {
	for _, tag := range summary.Tags {
		if strings.ToLower(strings.TrimSpace(tag)) == "autosave_snapshot" {
			return true
		}
	}
	return false
}

func latestSnapshotCreatedAt(snapshots []models.EngramSummary) *time.Time {
	if len(snapshots) == 0 {
		return nil
	}
	latest := snapshots[0].CreatedAt
	return &latest
}

func (run *sessionLifecycleRun) createAutosaveSnapshot() (*uuid.UUID, error) {
	messages, err := run.dependencies.ListChatMessages(
		run.ctx,
		run.session.SessionID,
		run.actorUserID,
		500,
		0,
	)
	if err != nil {
		return nil, err
	}
	if len(messages) == 0 {
		return nil, nil
	}
	abstract := DeriveChatSnapshotAbstract(messages, 320)
	if IsLowValueSnapshotAbstract(abstract) {
		return nil, nil
	}
	if DuplicateSnapshotExists(abstract, run.autosaveSnapshots) {
		return nil, nil
	}
	now := time.Now().UTC()
	threadID := fmt.Sprintf("chat-session:%s:autosave", run.session.SessionID)
	retrievalText := retrievalTextFromMessages(messages, 8)
	created, err := run.dependencies.CreateEngram(
		run.ctx,
		models.MemoryEngramCreate{
			ProjectID:               run.session.ProjectID,
			ThreadID:                &threadID,
			Title:                   fmt.Sprintf("%s Autosave %s", run.session.Title, now.Format("2006-01-02 15:04:05")),
			Abstract:                abstract,
			DetailedSummaryMarkdown: transcriptMarkdown(run.session, messages),
			Tags: append(
				[]string{},
				append(autosaveSnapshotTags, fmt.Sprintf("autosave_strategy:%s", run.session.AutosaveStrategy))...,
			),
			Keywords:        []string{"autosave", "snapshot", string(run.session.Provider), run.session.ModelID},
			VisibilityScope: string(run.session.VisibilityScope),
			SourceSessionID: &run.session.SessionID,
			RetrievalText:   &retrievalText,
		},
		run.embeddingDim,
		run.actorUserID,
		"chat.autosave_snapshot",
	)
	if err != nil {
		return nil, err
	}
	return &created.EngramID, nil
}

func transcriptMarkdown(session models.ChatSessionRecord, messages []models.ChatMessageRecord) string {
	lines := []string{
		"# Chat Session Snapshot: " + session.Title,
		"",
		"- Session ID: " + session.SessionID.String(),
		"- Provider/Model: " + string(session.Provider) + "/" + session.ModelID,
		"",
	}
	for _, message := range messages {
		lines = append(lines,
			"## "+strings.ToUpper(message.Role)+" ("+message.CreatedAt.Format(time.RFC3339)+")",
			normalizeEmptyMessage(message.ContentText),
			"",
		)
	}
	return strings.TrimSpace(strings.Join(lines, "\n"))
}

func normalizeEmptyMessage(contentText string) string {
	trimmed := strings.TrimSpace(contentText)
	if trimmed == "" {
		return "(empty)"
	}
	return trimmed
}

func retrievalTextFromMessages(messages []models.ChatMessageRecord, tailCount int) string {
	if tailCount <= 0 {
		return ""
	}
	start := len(messages) - tailCount
	if start < 0 {
		start = 0
	}
	parts := make([]string, 0)
	for _, message := range messages[start:] {
		trimmed := strings.TrimSpace(message.ContentText)
		if trimmed == "" {
			continue
		}
		parts = append(parts, trimmed)
	}
	return strings.Join(parts, " ")
}

func truncateText(value string, maxChars int) string {
	if len(value) <= maxChars {
		return value
	}
	if maxChars <= 3 {
		return value[:maxChars]
	}
	return strings.TrimRight(value[:maxChars-3], " ") + "..."
}

func normalizeMaxChars(maxChars int) int {
	if maxChars <= 0 {
		return 320
	}
	return maxChars
}
